code/modules/ruins/objects_and_mobs/generic_hazards/hazard_code/floor.dm

```diff
@@ -0,0 +1,129 @@
+/*
+floor hazards!
+
+ids don't work for slipping
+*/
+
+/obj/structure/hazard/floor
+	name = "floor hazard"
+	desc = "tell a maptainer if you see this. YOWCH!"
+	icon_state = "spikepit"
+	//these all(most) need to be not dense
+	density = FALSE
+	var/dealt_damage = 20
+	var/damage_type = BRUTE
+	//does damage to legs when walked over, or arms when bumped
+	var/contact_damage = FALSE
+	//does chunks of damage while stood on.
+	var/random_damage = FALSE
+
+	//slips!
+	var/slippery = FALSE
+	var/knockdown_time = 3 SECONDS
+	//no slip when walking makes walking not slip, slide makes players fly in the direction they were going.
+	var/slip_flags = NO_SLIP_WHEN_WALKING | SLIDE
+	var/paralyze_time = 0
+	//player drops their items
+	var/forcedrop = FALSE
+
+	//launches upwards using do_random_effect between random_min and random_max. also flattens players a little. Needs a icon_state-launch state to look right.
+	var/launcher = FALSE
+	//damage done by launcher, these are very crunchy and have a windup
+	var/launcher_damage = 90
+	//warning sent 1 second before the launcher does animation & damage
+	var/launch_warning = "The floor glows and begins to float!"
+
+/obj/structure/hazard/floor/do_random_effect()
+	if(launcher)
+		launch_sequence()
+	if(random_damage)
+		random_damage()
+
+/obj/structure/hazard/floor/Initialize()
+	if(contact_damage)
+		enter_activated = TRUE
+	if(launcher || random_damage)
+		random_effect = TRUE
+	. = ..()
+	if(slippery)
+		AddComponent(/datum/component/slippery, knockdown_time, slip_flags, _paralyze = paralyze_time, _force_drop = forcedrop)
+
+/obj/structure/hazard/floor/proc/launch_sequence()
+	visible_message(span_warning(launch_warning))
+	icon_state = initial(icon_state) + "-launch"
+	sleep(10)
```

Comment:
Try and avoid using sleep as its prone to causing overtime, I would swap this to an addtimer with a callback instead
something like 
```
addtimer(CALLBACK(src, PROC_REF(acctually_launch)), 0.5 SECONDS)
```